lighthouse-core/report/html/renderer/report-ui-features.js

```diff
@@ -391,8 +408,69 @@ class ReportUIFeatures {
    */
   onToolsButtonClick(e) {
     e.preventDefault();
-    this.toolsButton.classList.toggle('active');
-    this._document.addEventListener('keydown', this.onKeyDown);
+    e.stopImmediatePropagation();
+
+    if (this.toolsButton.classList.contains('active')) {
+      this.closeToolsDropdown();
+    } else {
+      this.openToolsDropdown();
+    }
+  }
+
+  /**
+   * Handler for tool button.
+   * @param {KeyboardEvent} e
+   */
+  onToolsButtonKeydown(e) {
+    switch (e.code) {
+      case 'ArrowUp':
+      case 'ArrowDown':
+      case 'Enter':
+      case ' ':
+        e.preventDefault();
+        this.openToolsDropdown();
+        break;
+      default:
+       // no op
+    }
+  }
+
+  /**
+   * Handler for tool dropdown.
+   * @param {KeyboardEvent} e
+   */
+  onToolsDropDownKeydown(e) {
+    const el = /** @type {?Element} */ (e.target);
+
+    if (!el) {
+      return;
+    }
+
+    let nextEl = el.nextElementSibling || this._dom.find(':first-child', this.toolsDropDown);
+    let prevEl = el.previousElementSibling || this._dom.find(':last-child', this.toolsDropDown);
+
+    if ((nextEl.getAttribute('disabled') !== null) ||
+     (window.getComputedStyle(nextEl).display === 'none')) {
+      nextEl = nextEl.nextElementSibling;
+    }
+
+    if ((nextEl.getAttribute('disabled') !== null) ||
+      (window.getComputedStyle(prevEl).display === 'none')) {
+      prevEl = prevEl.previousElementSibling;
+    }
+
+    switch (e.code) {
+      case 'ArrowUp':
+        e.preventDefault();
+        /** @type {HTMLElement} */ (prevEl).focus();
```

Comment:
thoughts on deferring calculating the "real" next/previous element to select until here, so only one is calculated?

Reply:
Sounds good to me; I didn't do this in my first pass because I ran into the [no-case-declaration lint error](https://eslint.org/docs/rules/no-case-declarations) but I see now that was because I forgot to wrap the case in { brackets }.

Reply:
Done, in my latest update I have moved all the work into helper functions called from within the switch cases.